## Lambdas - Define and Invoke

In [1]:
def square(x):
    
    return x * x

In [2]:
result = square(5)

result

25

In [3]:
square

<function __main__.square>

In [4]:
sq = square

In [5]:
sq

<function __main__.square>

In [6]:
result = sq(12)

result

144

In [8]:
cube_of = lambda x: x * x * x

In [9]:
result = cube_of(5)

result

125

In [10]:
result = cube_of(10)

result

1000

In [24]:
cube_of

<function __main__.<lambda>>

In [3]:
add = lambda x, y: x + y

In [13]:
sub = lambda x, y, z: x - y - z

In [5]:
mul = lambda x, y: x * y

In [6]:
div = lambda x, y: x / y

In [7]:
def calculate(x, y, operation="add"):
    
    if operation == "add":
        return add(x, y)
    if operation == "sub":
        return sub(x, y)
    if operation == "mul":
        return mul(x, y)
    if operation == "div":
        return div(x, y)

In [8]:
calculate(100, 4, operation="sub")

96

In [9]:
calculate(100, 4)

104

In [10]:
calculate(100, 4, operation="mul")

400

In [14]:
def calculate(*args, operation="add"):
    
    if operation == "add":
        return add(*args)
    if operation == "sub":
        return sub(*args)
    if operation == "mul":
        return mul(*args)
    if operation == "div":
        return div(*args)

In [15]:
calculate(100, 4, 5, operation="sub")

91

In [16]:
calculate(100, 3, operation="mul")

300

In [17]:
calculate(100, 3)

103

## Lambdas - Define, Invoke, and Discard

In [42]:
def calculate(*args, operation=add):
    
    return operation(*args)

In [43]:
calculate(100, 4, operation=sub)

TypeError: <lambda>() missing 1 required positional argument: 'z'

In [44]:
calculate(100, 4, 5,  operation=sub)

91

In [45]:
calculate(100, 4, operation=div)

25.0

In [46]:
calculate(10, 4, operation=lambda x, y: x**y)

10000

In [47]:
(lambda x: x + 2)(10)

12

In [48]:
(lambda x: x * 10)(22)

220

In [49]:
(lambda x, y: x ** y)(10, 2)

100

In [50]:
def check_if_even(x):
    
    assert x % 2 == 0

In [51]:
check_if_even(2)

In [29]:
check_if_even(3)

AssertionError: 

In [52]:
even_check_lambda = lambda x : assert x % 2 == 0

SyntaxError: invalid syntax (Temp/ipykernel_25652/1945895345.py, line 1)

In [53]:
total_score = lambda math, physics, chemistry, biology: math + physics + chemistry + biology

In [54]:
total_score(44, 56, 89, 92)

281

In [55]:
total_score(math=44, physics=56, chemistry=89, biology=92)

281

In [56]:
total_score = lambda math, physics, chemistry, biology=50: math + physics + chemistry + biology

In [35]:
total_score(44, 56, 89)

239

In [57]:
total_score(44, 56)

TypeError: <lambda>() missing 1 required positional argument: 'chemistry'

In [37]:
(lambda *numbers: sum(numbers))(3, 5, 6, 7)

21

In [58]:
(lambda *numbers: sum(numbers))(3, 5, 6, 7, 10)

31

In [59]:
(lambda **num_dictionary: sum(num_dictionary.values()))(a=23, b=34, c=45)

102

In [60]:
(lambda **num_dictionary: sum(num_dictionary.values()))(a=23, b=34)

57

## Lambdas - Filter() Function 

In [61]:
num_list = [1, 5, 6, 7, 11, 78, 99, 34, 105, 214]

In [62]:
filter(lambda x: x > 10, num_list)

In [63]:
greater_than_10_list = list(filter(lambda x: x > 10, num_list))

greater_than_10_list

[11, 78, 99, 34, 105, 214]

In [64]:
even_list = list(filter(lambda x: x % 2 == 0, num_list))

even_list

[6, 78, 34, 214]

In [65]:
greater_than_10_less_than_100_list = list(filter(lambda x: x > 10 and x < 100, num_list))

greater_than_10_less_than_100_list

[11, 78, 99, 34]

In [66]:
import text

ModuleNotFoundError: No module named 'text'

## Lambdas - Define and Invoke

Objectives
define and invoke lambda functions
[Video description begins] Topic title: Lambdas - Define and Invoke. Your host for this session is Janani Ravi. [Video description ends]

All of the functions that we've worked with so far have had names and we've invoked or called those functions using names.

[Video description begins] A Jupyter notebook titled Lambdas is displayed on the screen. It has a menu bar with the following options: File, Edit, View, etc. A set of code cells is present in the center pane. [Video description ends]

#### But what if a function doesn't have a name? Well, in that case, that function is a Lambda, and Lambdas are exactly what we're going to discuss here in this demo.

[Video description begins] Two code lines appear in code cell 1. Line 1 is: def square(x):. Line 2 is: return x * x. [Video description ends]

Let's work with a conventional function first. This is a function that we are familiar with which calculates the square of a number and this number is passed in as an input argument x. And the return value from this function is x multiplied by x, that is the square of the number. Once we have this custom function defined, we can use the name of the function to invoke the function. Here, we invoke the square function, pass in 5, and the result that we get stored in the result variable is 25.

[Video description begins] Two code lines appear in code cell 2. Line 1 is: result = square(5). Line 2 is: result. [Video description ends]

If you see what exactly this square name contains, you'll see that it's a function object. This is a function object defined in the main module.

[Video description begins] The following code appears in code cell 3: square. The output is: <function _main_.square(x)>. [Video description ends]

Don't worry about the main module or what it is yet. The main module is simply the default module in Python. We've seen that functions are first-class citizens in Python, which means that you can assign a function by name to another variable.

[Video description begins] The following code appears in code cell 4: sq = square. [Video description ends]

Here I've assigned the function square to the sq variable and you can see that sq is the same function object.

[Video description begins] The following code appears in code cell 5: sq. The output is: <function _main_.square(x)>. [Video description ends]

You can now calculate the square of the number by invoking this sq variable. Here, we invoke the square function using the sq variable.

[Video description begins] Two code lines appear in code cell 6. Line 1 is: result = sq(12). Line 2 is: result. [Video description ends]

And the result, which is the square of 12, is 144. Now, let's define a function with no name. This is an anonymous function or a lambda in Python.

[Video description begins] The following code appears in code cell 7: cube_of = lambda x: x * x * x. [Video description ends]

Observe that when you define a function with no name, you use the lambda keyword, not the def keyword. Here is a lambda function to calculate the cube of a number. Now, lambdas can get a little complicated to understand, so let's really pause for a bit and parse the syntax of this lambda. Focus your attention on just the right-hand side of this assignment statement.

We have the lambda keyword indicating we're defining an anonymous or nameless function. We then have x. x is the input argument to our lambda. In this particular example, x is the only input argument to our lambda. That is our anonymous function. Then we have the colon, after which we have an expression. There are certain constraints on lambdas in Python. You can't have complex statements and control structures such as if, else or for loops within lambdas.

You can only have an expression and the result of this expression is what is returned from this lambda. The input argument in our example is x, just the single input argument, and the expression is x multiplied by x multiplied by x, that is the cube of x. I've assigned this lambda function to a variable called cube_of and we'll use this cube_of variable to invoke this lambda function.

[Video description begins] Two code lines appear in code cell 8. Line 1 is: result = cube_of(5). Line 2 is: result. [Video description ends]

Let's invoke cube_of and pass in 5 as an input argument and the result that we get is 125. That is 5 cube. Observe how the expression within the lambda function is the return value of the lambda. Once you've assigned the lambda function to a variable, you can invoke it just like any other function.

[Video description begins] Two code lines appear in code cell 9. Line 1 is: result = cube_of(10). Line 2 is: result. [Video description ends]

Use the variable, invoke the function, pass in the input argument. The cube of the number 10 is, as we know, 1,000 and that is what is stored in the result variable. So what exactly is contained within this cube_of variable? Let's print out the details and see. You can see that it is a function object. And you can see from its specification that it's a lambda function, it's not a named function.

[Video description begins] The following code appears in code cell 10: cube_of. The output is: <function _main_.<lambda>(x)>. [Video description ends]

I'm going to go back to the calculator example that we worked with before. But this time I'm going to use lambdas to define the various arithmetic operations.

[Video description begins] The following code appears in code cell 11: add = lambda x, y: x + y. [Video description ends]

Once again, I use the lambda keyword to define my lambda, and observe that this lambda takes in two input arguments, x and y. The input arguments are followed by the colon and the expression here is x + y, x + y is what will be returned as a result of executing this lambda. I'll now define a lambda function for the subtraction operation as well. This lambda once again takes in two input arguments, x and y, and the expression is x- y. That's what will be returned.

[Video description begins] The following code appears in code cell 12: sub = lambda x, y: x - y. [Video description ends]

Here is a lambda for the multiplication operation assigned to the mul variable.

[Video description begins] The following code appears in code cell 13: mul = lambda x, y: x * y. [Video description ends]

x and y are the input arguments, x multiplied by y is the expression. And finally, here is the lambda for the division operation.

[Video description begins] The following code appears in code cell 14: div = lambda x, y: x / y. [Video description ends]

x, y are the input arguments, x divided by y is the expression which is the result that will be returned. We can now define a custom function called calculate, which will perform any arithmetic operation that you specify. And this time we'll have this calculate function invoke lambdas.

[Video description begins] Nine code lines appear in code cell 15. Line 1 is: def calculate(x, y, operator=''add''):. Line 2 is: if operator == "add":. Line 3 is: return add(x, y). Line 4 is: if operator == "sub":. Line 5 is: return sub(x, y). Line 6 is: if operator == "mul":. Line 7 is: return mul(x, y). Line 8 is: if operator == "div":. Line 9 is: return div(x, y). A screenshot of code cells 11-14 appears. [Video description ends]

This calculate function takes in three input arguments, x, y and the operation to be performed using x and y. If the operation is add then we invoke the add lambda and pass in x and y to perform the addition. If operation is sub, we invoke the sub lambda to perform the subtraction operation. And if the operation specified is mul or div, we invoke the corresponding mul and div lambdas. Let's confirm that our calculate function using lambdas does indeed work.

[Video description begins] The following code appears in code cell 16: calculate(100, 4, operator=''sub''). [Video description ends]

Here I've invoked the calculate function, x and y are 100 and 4 respectively, and the operation that I want to perform is the subtraction operation. And I get the result here on my screen, 96, 100 minus 4 is equal to 96. If I invoke the calculate function, only specifying values for x and y and not for the operation. The default operation that will be performed will be the addition operation by invoking the add lambda. 100 plus 4 as we know is 104.

[Video description begins] The following code appears in code cell 17: calculate(100, 4). The output is 104. [Video description ends]

Let's invoke the calculate function once again to perform the multiplication operation on the same two x and y values, 100 and 4. 100 multiplied by 4 gives us 400.

[Video description begins] The following code appears in code cell 18: calculate(100, 4, operator=''mul''). The output is 400. [Video description ends]

And just like in the last video that we looked at, we can improve this calculate function by taking in variable length arguments using *args.

[Video description begins] Nine code lines appear in code cell 19. Line 1 is: def calculate(*args, operator=''add''):. Line 2 is: if operator == "add":. Line 3 is: return add(*args). Line 4 is: if operator == "sub":. Line 5 is: return sub(*args). Line 6 is: if operator == "mul":. Line 7 is: return mul(*args). Line 8 is: if operator == "div":. Line 9 is: return div(*args). [Video description ends]

This is the exact same calculate function as before where we invoke lambdas to perform the arithmetic operations of addition, subtraction, multiplication, and division. Instead of using specific arguments for x and y values, we use the variable length argument named args. Whatever argument is passed in is stored in the args tuple within this function. And we unpack this args tuple when we invoke our lambda, whether it's add, sub, mul, or div. We'll improve our calculate function further in just a bit. But before that, let's test this out and make sure that it indeed does work.

[Video description begins] The following code appears in code cell 20: calculate(100, 4, operator=''sub''). [Video description ends]

Here is my function invocation to perform subtraction. 100 minus 4 gives us 96.

[Video description begins] The following code appears in code cell 21: calculate(100, 3, operator=''mul''). [Video description ends]

Here is my function invocation to perform multiplication. 100 multiplied by 3 gives us 300. And finally, if you invoke the calculate function and don't specify an operation, it will add your input arguments. 100 plus 3 is equal to 103 that's because the default specification for the operation is add.

[Video description begins] The following code appears in code cell 22: calculate(100, 3). The output is 103. 

## Lambdas - Define, Invoke, and Discard

Objectives
define, invoke, and discard lambdas
[Video description begins] Topic title: Lambdas - Define, Invoke, and Discard. Your host for this session is Janani Ravi. [Video description ends]

#### We know that functions are first-class citizens in Python, which means you can pass in functions as an input argument to another function.

[Video description begins] A Jupyter notebook titled Lambdas is displayed on the screen. It has a menu bar with the following options: File, Edit, View, etc. A set of code cells is present in the center pane. [Video description ends]

#### But are lambdas first-class citizens as well? Yes, indeed they are, lambdas are just functions with no name. We'll see how lambdas can be passed in as an input argument here in this demo.

[Video description begins] Two code lines appear in code cell 23. Line 1 is: def calculate(*args, operation=add):. Line 2 is: return operation (*args). [Video description ends]

I'm going to redefine the calculate function that we've worked with so far. The first input argument here is *args, which is a variable length argument. This is what we'll use to pass in the numbers on which we want the calculation to be performed. The second input argument to this function is the function object representing the calculation that we want to perform.

This is the function object that will actually perform the calculation. The function object is stored in the operation variable. And we invoke this function within this calculate function body, and pass in *args as an input to this operation function. Why *args and not args directly? Well, we've discussed this earlier, *args will unpack the args tuple into its individual elements. And those individual elements will be passed into the operation function. I'm now going to invoke this calculate function.

[Video description begins] The following code appears in code cell 24: calculate(100, 4, operation=sub). A screenshot of code cells 11-14 from the previous video appears on the screen. [Video description ends]

The input arguments, that is the numbers on which I want the calculation to be performed are 100 and 4. And the arithmetic operation that I want to perform is the subtraction operation. So I pass in the sub lambda as my function object. As we know, the sub lambda that we have referenced here subtracts y from x giving us 96, 100 minus 4 is equal to 96. If you want to perform the division operation on two numbers, you'll simply change the lambda that you passed in as the operation input argument.

[Video description begins] The following code appears in code cell 25: calculate(100, 4, operation=div). [Video description ends]

You specified 100, 4 and the operation that we want to perform is the div, and 100 divided by 4 gives us 25. Now so far, we always assigned lambdas to variables, and then invoked lambdas using the name of the variables.

[Video description begins] The following code appears in code cell 26: calculate(10, 4, operation=lambda x, y: x**y). [Video description ends]

But the whole idea behind using lambdas is that you can whip them up on the fly, use it when you need it, and then discard it. Observe that the operation input argument here is a lambda specification directly. This lambda takes in two input arguments x and y, and calculates x to the power y, it's a power of lambda. This lambda is invoked on the input arguments that we have specified, giving us 10 raised to the power of 4 which is 10,000. You can see that when you want to use lambdas for one off operations, operations that you want to reuse, you can simply create these lambdas on the fly.

[Video description begins] The following code appears in code cell 27: (lambda x: x+2) (10). [Video description ends]

Here is a seemingly complicated bit of code which uses a lambda. Let's focus on the lambda definition first. This takes in one input argument, x adds 2 to x, and that is the return value from this lambda. This statement is complicated because I've defined the lambda function within parenthesis and then invoked it directly. And the input argument for my invocation is 10. This lambda will increment 10 by 2 and the result will be 12, as you can see here on screen. And when you see lambdas used in this way, that's when you realize how lambdas are use and throw functions, anonymous functions with no names which can be used and then discarded.

[Video description begins] The following code appears in code cell 28: (lambda x: x*10) (22). [Video description ends]

Here is a lambda definition that multiplies the input x by 10, and that is the return value from this lambda. And I'm going to invoke this lambda right away on 22, and the result is 22 multiplied by 10 equal to 220. Here is another anonymous lambda that I'm going to define and invoke right away. It takes in two input arguments x and y, and calculates x to the power y.

[Video description begins] The following code appears in code cell 29: (lambda x, y: x**y) (10, 2). [Video description ends]

I invoke this lambda right away on 10 and 2, it calculates 10 to the power 2, giving us the result 100. Let's go back to a regular function for just a bit in order to understand the constraints that you have when you use lambdas.

[Video description begins] Two code lines appear in code cell 30. Line 1 is: def check_if_even(x):. Line 2 is: assert x % 2 == 0. [Video description ends]

Here is a function definition which checks whether the input number that is passed in is even. This takes in one input argument, x, and it uses an assert statement. Assert checks whether x modulo 2 is equal to 0. The way this assert statement works is that it checks whether x is even. If it is, it does nothing. If x is not even, that is, x modulo 2 is not equal to 0, it will raise an error. Let's try this out, I call check_if_even and pass in 2. This is indeed even, and you can see that it returns nothing, assert does nothing.

[Video description begins] He points at code cell 31. It has the following code: check_if_even(2). The following code appears in code cell 32: check_if_even(3). [Video description ends]

But if you call check_if_even and pass in 3, you'll see that there is an AssertionError. You can see that the AssertionError says that the assert was not satisfied, x is not even. check_if_even is a fairly simple function. Let's see if we can define the same function using a lambda.

[Video description begins] The following code appears in code cell 33: even_check_lambda = lambda x : assert x % 2 == 0. [Video description ends]

This lambda takes in x, that is our single input argument, and within the lambda, the expression is assert x % 2 == 0. And when you try to execute this, you'll find that you'll get an error. This syntax is invalid, you cannot have keywords and other statements within your lambda. The body of a lambda can only contain expressions. You can have lambdas with as many input arguments as you want.

[Video description begins] The following code appears in code cell 34: total_score = lambda math, physics, chemistry, biology: math + physics + chemistry + biology. [Video description ends]

Here is a lambda that takes in four scores, math, physics, chemistry, and biology, and calculates the total score by summing up all of the individual scores. This lambda is assigned to the total score variable, which is what I invoke to calculate the total which happens to be 281.

[Video description begins] The following code appears in code cell 35: total_score(44, 56, 89, 92). The output is: 281. [Video description ends]

So far, we've always invoked lambdas by using positional arguments.

[Video description begins] The following code appears in code cell 36: total_score(math=44, physics=56, chemistry=89, biology=92). [Video description ends]

Just like ordinary named functions, lambdas can be invoked using keyword arguments as well, as you can see here in this example. Math score is 44, physics 56, chemistry 89, and biology 92. Using keyword arguments, as we already know, is much easier to read and makes our code more maintainable.

[Video description begins] The following value appears for code cell 36: 281. [Video description ends]

You can also have your lambdas specify default values for their argument.

[Video description begins] The following code appears in code cell 37: total_score = lambda math, physics, chemistry, biology=50: math + physics + chemistry + biology. [Video description ends]

Here is a lambda which calculates the total score, observe that we have specified a default biology score of 50. You can invoke this lambda by leaving out the biology score.

[Video description begins] The following code appears in code cell 38: total_score(44, 56, 89). [Video description ends]

I've specified just three scores here, they'll be associated with math, physics, and chemistry. The lambda will consider the default score for biology which is 50, giving us a total score of 239. Just like with ordinary functions, if you try to invoke a lambda and specify fewer values than number of arguments, that is an error. And the error here clearly says that our lambda is missing one required positional argument.

[Video description begins] The following code appears in code cell 39: total_score(44, 56). A TypeError appears in the output. It reads: <lambda>() missing 1 required positional argument: 'chemistry'. [Video description ends]

Lambdas can also specify a variable length input arguments.

[Video description begins] The following code appears in code cell 40: (lambda *numbers: sum(numbers))(3, 5, 6, 7). [Video description ends]

Observe our lambda definition here, you can see that its input argument is *numbers. And within the lambda, our expression simply invokes the sum function. The sum function passes into its input the numbers tuple to get the sum of all of the numbers that we've passed in. And just for fun, in order to make things more complicated and make our code harder to read, I've defined the lambda and invoked it immediately on the numbers 3, 5, 6, and 7. When you execute this code, you'll get the sum of all of these numbers which is equal to 21. Remember that lambdas which are defined and invoked right away are use and throw functions

[Video description begins] The following code appears in code cell 41: (lambda *numbers: sum(numbers))(3, 5, 6, 7, 10). [Video description ends]

which means that if you want to perform the same sum operational once again using a lambda, you need to define the lambda and then invoke it as you see here on screen. Invoking this lambda on 3, 5, 6, 7, 10, gives us the sum 31. You can also define your lambdas and have them take in variable length keyword arguments. Here is a lambda which takes in a variable length keyword argument called num_dictionary.

[Video description begins] The following code appears in code cell 42: (lambda **num_dictionary: sum(num_dictionary.values() )) (a=23, b=34 , c=45). [Video description ends]

You know this because we have the double asterisk before num_dictionary. The expression within this lambda simply accesses the values property of the dictionary and calculates the sum of values. We've defined this lambda and invoked it right away using keyword arguments, a=23, b=34, and c=45. And the sum of these three numbers is equal to a 102. Let's define and invoke the same lambda once again, and this time we've specified two keyword arguments, a=23, b=34, giving us the sum 57.

[Video description begins] The following code appears in code cell 43: (lambda **num_dictionary: sum(num_dictionary.values() )) (a=23, b=34). 

## Lambdas - Filter() Function 

Objectives
use lamdas as an input to the built-in filter() function
[Video description begins] Topic title: Lambdas - Filter() Function. Your host for this session is Janani Ravi. [Video description ends]

We've worked with Lambdas a fair bit so far. We've seen that lambda functions can get fairly complicated, moreover, they are hard to read. So when would you choose to use lambdas rather than regular functions?

[Video description begins] A Jupyter notebook titled Lambdas is displayed on the screen. It has a menu bar with the following options: File, Edit, View, etc. A set of code cells is present in the center pane. [Video description ends]

Lambdas are often used along with the built-in filter function in Python.

[Video description begins] The following code appears in code cell 44: num_list = [1, 5, 6, 7, 11, 78, 99, 34, 105, 214]. [Video description ends]

In order to understand how this filter function works, I'm going to set up a list. This is a simple list of numbers, it contains odd numbers and even numbers, small numbers, very large numbers. Let's invoke the filter function. The filter function takes in two input arguments.

[Video description begins] The following code appears in code cell 45: filter(lamda x: x>10, num_list). [Video description ends]

The first of these is a function object. And this function object is often specified using a lambda function as you see here on screen. The second input argument is the list that you want to filter. Let's take a look at the first input argument here that is the lambda function. This lambda takes in one input argument x, and it checks to see whether x is greater than 10.

The expression x greater than 10 will return true if x is greater than 10, will return false otherwise. The way the filter function works is that this lambda is applied to every element of num_list. Based on whether a particular element in num_list is greater than or less than 10, this lambda will return true or false. The filter function will then retain all of those elements where the lambda returns true. Don't worry if this seems confusing at first, a few examples will make this very clear. The filter function returns a filter object, and a filter object is not something that you can understand right away.

[Video description begins] An output appears below code cell 45. It reads: <filter at 0x112e167b8>. [Video description ends]

However, a filter object can be converted to a list and that's exactly what we'll do. Here is how we would typically use the filter function in the real world.

[Video description begins] Two lines of code appear in code cell 46. Line 1 is: greater_than_10_list = list(filter(lambda x: x >10, num_list)). Line 2 is: greater_than_10_list. [Video description ends]

We have our lambda definition which checks to see whether x is greater than 10. This lambda is applied to num_list, so the filter function will apply this lambda function to every element of num_list. For each of those elements where the condition returns true, that is, the number is greater than 10, that will be returned. So we form a list of those elements where x is greater than 10, and that is what will be stored in the variable greater_than_10_list. And you can see that this is indeed true by looking at the result here.

[Video description begins] The following output appears: [11, 78, 99, 34, 105, 214]. [Video description ends]

The greater_than_10_list contains all of those numbers from our original num_list, which were greater than 10. Let's make sure that we really understand how filter works by taking a look at another example. A lambda function is the most important and that's what we look at first.

[Video description begins] Two lines of code appear in code cell 47. Line 1 is: even_list = list(filter(lambda x: x % 2 == 0, num_list)). Line 2 is: even_list. [Video description ends]

A lambda takes in one input argument x, and checks to see whether x % 2 == 0. It checks to see whether x is even. If x is even, that is x modulo 2 is equal to 0, the lambda will return true. This lambda is applied to every element of num_list and the filter function will return all of the even elements which is stored in the even_list variable. Remember to pass in the filter object into the list function to get the result in the form of a list.

[Video description begins] The following output appears: [6, 78, 34, 214]. [Video description ends]

One last filter function before we move on from lambdas to other topics.

[Video description begins] Two lines of code appear in code cell 48. Line 1 is: greater_than_10_less_than_100_list = list(filter(lambda x: x >10 and x < 100, num_list)). Line 2 is: greater_than_10_less_than_100_list. [Video description ends]

Here is a lambda function which checks to see whether x is greater than 10 and less than 100. There are two conditions that we're checking for within the lambda expression and this is allowed. This lambda function is applied to every element of num_list, so as a result of this filter function and the list operation, we get all numbers which are greater than 10 but less than 100 in our result.